BasicSample/Assets/Interaction/Scripts/HapticsSupport.cs

```diff
@@ -0,0 +1,29 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.XR;
+
+namespace Microsoft.MixedReality.OpenXR.BasicSample
+{
+    public class HapticsSupport : MonoBehaviour
+    {
+        private readonly List<InputDevice> controllerInputDevices = new List<InputDevice>();
+
+        // Update is called once per frame
+        void Update()
+        {
+            InputDevices.GetDevicesWithCharacteristics(InputDeviceCharacteristics.Controller, controllerInputDevices);
+
+            foreach (InputDevice controller in controllerInputDevices)
```

Comment:
@keveleigh  is this the best practice to loop through the "input device" collection and check the capabilities every frame?  Is there a way to listen to the input device changes and pre-compute the capabilities without recompute it in every "Update" function?

Reply:
You're right that it wouldn't avoid the loop to query `TryGetFeatureValue` (unless you decide to do something like explicitly cache one `leftController` and one `rightController`, but I don't think that's necessary here).

The nice thing is that you'd be able to avoid the `InputDevices.GetDevicesWithCharacteristics` (which we've found _can_ be expensive every frame) and you can also probably avoid the `if (controller.TryGetHapticCapabilities(out HapticCapabilities hapticCapabilities) && hapticCapabilities.supportsImpulse)` check by confirming that once in the `deviceConnected` callback instead.

Reply:
Done